model/DEL_network.py

```diff
@@ -0,0 +1,278 @@
+        self.model_layers.append(torch.nn.Linear(hidden_list[-1], 6))
+
+    def forward(self, q, q_next, h):
+        r1 = q[0:3]
+        Q1 = q[3:7]
+        r2 = q_next[0:3]
+        Q2 = q_next[3:7]
+        omega = 2 / h * (H.T @ L(Q1).T @ Q2)
+        Q_mid = Q1 + 0.5 * h * (0.5 * L(Q1) @ H @ omega)  # TODO: Determine if this is the best implementation
+
+        r_mid = (r1 + r2) / 2
+        r_mid_vel = (r2 - r1) / h
+
+        out = torch.cat((r_mid, Q_mid, r_mid_vel, omega))
+        for layer in self.model_layers:
+            out = layer(out)
+
+        return out
+
+
+class ControlInputJacobianNetwork(torch.nn.Module):
+    def __init__(self, D_in, num_control_inputs, hidden_list):
+        """
+        Neural Network used to represent the control input Jacobian for the system
+        """
+        super(ControlInputJacobianNetwork, self).__init__()
+        self.hyperparams = None
+        self.num_control_inputs = num_control_inputs
+        self.model_layers = torch.nn.ModuleList()
+
+        # input layer
+        self.model_layers.append(torch.nn.Linear(D_in, hidden_list[0]))
+        self.model_layers.append(torch.nn.Dropout())
+        self.model_layers.append(torch.nn.Tanh())
+        # add all hidden layers
+        for i in range(0, len(hidden_list) - 1):
+            self.model_layers.append(torch.nn.Linear(hidden_list[i], hidden_list[i + 1]))
+            self.model_layers.append(torch.nn.Dropout())
+            self.model_layers.append(torch.nn.Tanh())
+
+        # output layer
+        self.model_layers.append(torch.nn.Dropout())
+        self.model_layers.append(torch.nn.Linear(hidden_list[-1], 6*num_control_inputs))
+
+    def forward(self, q, q_next, h):
+        """
+        XXXX
+        """
+        r1 = q[0:3]
+        Q1 = q[3:7]
+        r2 = q_next[0:3]
+        Q2 = q_next[3:7]
+        omega = 2 / h * (H.T @ L(Q1).T @ Q2)
+        Q_mid = Q1 + 0.5 * h * (0.5 * L(Q1) @ H @ omega) # TODO: Determine if this is the best implementation
+
+        r_mid = (r1 + r2) / 2
+        q_mid = torch.cat((r_mid, Q_mid))
+
+        out = q_mid
+
+        for layer in self.model_layers:
+            out = layer(out)
+
+        return torch.reshape(out, (6, self.num_control_inputs))
+
+class DELNetwork(torch.nn.Module):
+    def __init__(self, num_states):
+        """
+        Neural Network used to approximate the discrete Euler-Lagrange of the system
+        """
+        super(DELNetwork, self).__init__()
+        num_pose_states = 7
+        num_control_inputs = 4
+        self.h = 1/400  # Time Step of Dataset
+        hidden_list = [64, 64, 64, 64]
+
+        self.massMatrix = MassMatrixNetwork(num_pose_states, hidden_list)
+        self.potentialEnergy = PotentialEnergyNetwork(num_pose_states, hidden_list)
+        self.dissipativeForces = DissipativeForceNetwork(num_states, hidden_list)
+        self.controlJacobian = ControlInputJacobianNetwork(num_pose_states, num_control_inputs, hidden_list)
```

Comment:
should we include velocities in input jacobian? I would think RPMs will directly actuate velocities, and indirectly actuate positions